### PR TITLE
feat(typespec): upgrade to typespec 0.45.0

### LINKED ISSUE #3477: feat(typespec): upgrade to typespec 0.45.0
## Description

See release notes:
- TypeSpec Core release notes: [June 2023 | TypeSpec (microsoft.github.io)](https://microsoft.github.io/typespec/release-notes/release-2023-06-06)
- TypeSpec Azure release notes: [June 2023 | TypeSpec Azure](https://azure.github.io/typespec-azure/docs/release-notes/release-2023-06-06)

This is a blocker for some of our work items, like #2988

### PR DESCRIPTION
**Pull request by archerzz**
# Description

- upgrade typespec libraries to version 0.45.0
- upgrade `cadl-ranch` to 0.16.0
- fix regression due to breaking change in `tspconfig.yaml` resolution logic
- fix regression in `@parentResource` doesn't accept template parameter
- re-generate tests

resolve #3477

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/TypeSpec.Extension/Emitter.Csharp/package.json [hunk 1]
```diff
@@ -50,17 +50,17 @@
...
-        "@typespec/eslint-plugin": "0.44.0",
-        "@typespec/http": "0.44.0",
-        "@typespec/library-linter": "0.44.0",
-        "@typespec/openapi": "0.44.0",
-        "@typespec/rest": "0.44.0",
-        "@typespec/versioning": "0.44.0",
+        "@typespec/eslint-plugin": "0.45.0",
+        "@typespec/http": "0.45.0",
+        "@typespec/library-linter": "0.45.0",
+        "@typespec/openapi": "0.45.0",
```
Comment (chunyu3) [inline src/TypeSpec.Extension/Emitter.Csharp/package.json:61]: Please remove `@typespec/openapi`  and `@azure-tools/typespec-autorest` aslo. Thanks

File: test/CadlRanchProjects/azure/core/basic/src/Generated/tspCodeModel.json [hunk 2]
```diff
@@ -121,7 +121,7 @@
      "Description": "The entity tag for this resource.",
      "Type": {
       "$id": "17",
-      "Name": "string",
+      "Name": "eTag",
```
Comment (lirenhe) [inline test/CadlRanchProjects/azure/core/basic/src/Generated/tspCodeModel.json:124]: I think @m-nash talked about seperating typespec version upgrade with cadl-ranch upgrade.
I heard there is no breaking change in this typespec version, so I would like to confirm, if we didn't do anything on cadl-ranch, will all existing cases still work for new version of the emitter?

File: test/TestProjects/Authoring-TypeSpec/src/Generated/tspCodeModel.json [hunk 3]
```diff
@@ -670,7 +670,7 @@
             "Kind": "String",
             "IsNullable": false
            },
-           "IsRequired": true,
+           "IsRequired": false,
```
Comment (pshao25) [inline test/TestProjects/Authoring-TypeSpec/src/Generated/tspCodeModel.json:673]: How did this change happen?

File: eng/Generate.ps1 [hunk 4]
```diff
@@ -64,11 +64,17 @@ function Add-TestServer-Swagger ([string]$testName, [string]$projectSuffix, [str
 
 function Add-CadlRanch-TypeSpec([string]$testName, [string]$projectPrefix, [string]$cadlRanchProjectsDirectory) {
     $projectDirectory = Join-Path $cadlRanchProjectsDirectory $testName
+    $configFile = Join-Path $projectDirectory "tspconfig.yaml"
```
[the remark this answers is not in the document]
Comment (archerzz) [reply eng/Generate.ps1:67]: Yes. It's due to a breaking change in `0.45.0`.

**Review by pshao25 — APPROVED** [review]

File: eng/Generate.ps1 [hunk 5]
```diff
@@ -64,11 +64,17 @@ function Add-TestServer-Swagger ([string]$testName, [string]$projectSuffix, [str
...
     if (Test-Path "$projectDirectory/*.sln") {
         $projectDirectory = Join-Path $projectDirectory "src"
     }
     $tspMain = Join-Path $cadlRanchFilePath $testName "main.tsp"
-    Add-TypeSpec "$projectPrefix$testName" $projectDirectory $tspMain
+    if (Test-Path $configFile) {
+        Add-Typespec "$projectPrefix$testName" $projectDirectory $tspMain "--config=$configFile"
```
Comment (m-nash) [inline eng/Generate.ps1:73]: why is this needed, doesn't tsp look for the tspconfig.yaml in the same spot so it should automatically find this?

File: test/CadlRanchMockApis/package.json [hunk 6]
```diff
@@ -32,6 +32,6 @@
         "dist/**"
     ],
     "peerDependencies": {
-        "@azure-tools/cadl-ranch-specs": "0.15.0"
+        "@azure-tools/cadl-ranch-specs": "0.16.0"
```
Comment (m-nash) [inline test/CadlRanchMockApis/package.json:35]: do we absolutely need to take both a cadl ranch bump and tsp bump in the same PR?  We discussed this last time and it can make it difficult to understand which changes in this PR are due to tsp 45 and which changes are due to cadlranch 16.

File: test/CadlRanchProjects/azure/core/basic/src/Generated/tspCodeModel.json (hunk 2 above)
Comment (m-nash) [reply test/CadlRanchProjects/azure/core/basic/src/Generated/tspCodeModel.json:124]: there are many breaking changes in this typespec version so not sure what this comment means.

File: eng/Generate.ps1 (hunk 5 above)
Comment (archerzz) [reply eng/Generate.ps1:73]: @m-nash Some test projects have separate configuration, like https://github.com/Azure/autorest.csharp/blob/feature/v3/test/CadlRanchProjects/authentication/api-key/tspconfig.yaml

Those configuration files are not saved along with the typespec files. In `0.45.0`, there is a [breaking change](https://github.com/microsoft/typespec/pull/1942) on how typespec cli resolves the path of configuration file. So, we need to explicitly specify the path to the configuration file in our cases.

Comment (pshao25) [reply eng/Generate.ps1:73]: One more thing, typespec team thinks it is a bug fix, so it is not in the breaking change section of release notes. I don't really agree with them.

Comment (m-nash) [reply eng/Generate.ps1:73]: yeah, bug fixes can be breaking changes as well @pshao25.  @allenjzhang does tsp team consider bug fixes that are breaking changes as well should go into the breaking change section?

Comment (m-nash) [reply eng/Generate.ps1:73]: @archerzz  is this because the tsp is coming from cadl-ranch so the tspconfig.yaml is not next to the main.tsp?

File: test/CadlRanchMockApis/package.json (hunk 6 above)
Comment (m-nash) [reply test/CadlRanchMockApis/package.json:35]: We should probably push a change in cadl-ranch setting `@typespec/compiler": "^0.44.0"` so that we don't have to do this, then create this PR as 45 bump only.

Comment (m-nash) [reply test/CadlRanchMockApis/package.json:35]: Talked with the dpg team we probably need to do this as the next step.  https://github.com/Azure/cadl-ranch/issues/356

File: eng/Generate.ps1 (hunk 5 above)
Comment (archerzz) [reply eng/Generate.ps1:73]: @m-nash That's correct. All `tspconfig.yaml` files under https://github.com/Azure/autorest.csharp/tree/feature/v3/test/CadlRanchProjects are checked in our repo, not `cadl-ranch`.

**Comment by archerzz** [comment]
Regen PR: https://github.com/Azure/azure-sdk-for-net/pull/37064

There is no change on the `azure-sdk-for-net`.

**Review by lirenhe — APPROVED** [review]
